### ISSUE TITLE
Incorrect behavior of strict vs. default modes

### ISSUE THREAD

**Issue by pdlug** [ruby-json-schema/json-schema#103]
The behavior of strict mode is a bit misleading (and incorrect IMHO). JSON schema provides `required` already but in the default execution of the json-schema validator it ignores this property. The expected behavior is that default validation checks that any keys present in the document being validated match the definition in the schema and that all keys listed as `required` in the schema are present. Strict mode would then make a bit more sense to change all keys to required though I'd argue that's a bit misleading as well and strict mode should really just imply that there are no keys present in the validation target that are not defined in the schema as well. Specifying that all keys present in the schema are required really should be a separate option since this is a different validation case (perhaps `require_all: true`?).

In summary I propose:

Default validation:
- Invalid unless all keys in the input document match the definition in the schema.
- Invalid unless all keys listed as required in the schema are present. 
- Keys present in the input document but not present in the schema are ignored (i.e don't affect validation).

Strict validation:
- Default rules + invalid if any keys are present in the input document but not defined in the schema.

Require all:
- Strict validation + invalid unless all keys defined in the schema are present in the input document.

I'm happy to take a stab at a patch if I can get some consensus that this is the right thing to do.

**Comment by tristil** [comment]
FWIW there's a [thoughtbot post]( https://robots.thoughtbot.com/validating-json-schemas-with-an-rspec-matcher) where they use `strict: true` so if people copy that matcher they may lose some time trying to figure out why for example the parser isn't respecting `required` lists, etc. I don't really get what strict is for either, it ends up just distorting the schema you've specified.

**Comment by freiden** [comment]
Sorry for being late to the party but is strict option is still needed since behaviors are defined in schemas?

**Comment by RST-J** [comment]
@freiden What exactly do you mean? Strict is a makro so to say which prohibits any properties not covered by the schema (default in JSON schema is to accept any properties about which nothing is stated in a schema) and at the same time make every property required which is listed in a schema (properties not explicitly stated as required can be missing by default).

It is not necessary to have this as one can achieve this by hand. However, it was introduced as shortcut so one has not to do this tedious steps to convert relaxed schemas to strict ones, iirc.

**Comment by freiden** [comment]
If I understand well the purpose is to validate schemas based on the specfication, so validations of fields should be managed by the schema to prevent unwanted behavior in context where you doesn't use this gem for validations. 
For example, I'm using the gem to validate my schema during tests but in my API I don't need the validation when sending my data the rest of the time.

**Comment by cec** [comment]
@iainbeeston correct me if I'm wrong, but as stated [here](http://json-schema.org/example2.html) under the base schema section, `additionalProperties` is specified at the object level, not for every property; it would be a nightmare otherwise. The same is stated in the [Understanding JSON Schema](http://spacetelescope.github.io/understanding-json-schema/reference/object.html#properties) "web book" (I know there's a better name for this...)

**Comment by iainbeeston** [comment]
@cec That's correct. If you're referring to my comment above - that's a typo (should have said for every object, or maybe "for every property")

**Comment by cec** [comment]
@iainbeeston I'd say *for every object*. It's a very property anyway :)

**Comment by cec** [comment]
@iainbeeston @RST-J I recently began a new project where an extensive use of combining schemas is required and this made me realise that @pdlug proposal is definitely necessary.
Allow me to explain.

We use JSON Schema to describe responses of our APIs endpoints, similar to specifying a protocol.
When doing so, it is of utmost importance to be able to programmatically ensure that API clients do not receive unexpected data.

`"additionalProperties" : false` guarantees that the protocol is not violated unintentionally when changed the API.

However, specifying this prevents from combining/extending schemas using `allOf` (see [this issue on the json-schema repo](https://github.com/json-schema/json-schema/issues/116)).

The "ban unknown props mode" discussed on the referenced issue and in [the proposal for draft 5](https://github.com/json-schema/json-schema/wiki/ban-unknown-properties-mode-(v5-proposal)) represents the change proposed by @pdlug and by @pkarman in #295 .

I think efforts should be made to implement and merge this because of the following reasons:

- allowing props non mentioned in the scheme defeats one of the purposes of JSON Schema

- preventing these props with `additionalProperties:false` would require every schema extension/combination to be a complete copy of the schema being extended, with the additional properties defined right after

I apologize for the long post and look forward to your comments.

**Comment by Haelle** [comment]
I was thinking of a slightly different idea ; 
When `strict: true` offer the possibility to the users to "override" the default behavior enforced by the strict mode

Example :
with `strict: true` all property are by default `"required": true` ; but if the user set `"optional": true` or `"required": false` if override the behavior.
The same if the user set `"additionalProperties": true`

I could try and dig in to the code to see if I can make a PR out of this idea

WDYT ?

**Comment by AlexandreBernard** [comment]
I need exactly the same behavior as you do @Haelle 

I want strict validations because I want to validate that a json matches the schema and I don't trust the ones who are going to be adding fields to that JSON. So strict validation is great as it enforces everything to be defined in the schema

However, an optional attribute must be possible even in strict mode if defined in the schema. So if in the schema we have a required array that does not contain the attribute it should not be required even in strict mode.

`optional: true` does make sense but in non json schema standard.

`required: true/false` leads to problems as required should be a list of properties, not a boolean

**Comment by elik-ru** [comment]
Hey people!
Do i get it correctly that 10 year later there is sill no solution? Is there any objection against respecting `"required": false` in strict mode?

**Comment by elik-ru** [comment]
There is was a PR that was doing exactly this, but it did not receive any comments: https://github.com/voxpupuli/json-schema/pull/406 and closed by author because of silence.

**Comment by Drowze** [comment]
I've been following this issue for far too long... Hope this gets fixed someday. As of now it is simply not possible to define optional properties when using the strict mode.

For a change, this is what I'd expect under strict mode:

1. When the schema defines `required`, the validator should respect it.
```jsonc
// the validator should respect the "required" definition and allow "last_name" to be optional
{
  "type": "object",
  "properties": {
    "name": { "type": "string" },
    "last_name": { "type": "string" }
  },
  "required": ["name"]
}

// the validator should respect the "required" definition and allow any defined property to be optional:
{
  "type": "object",
  "properties": {
    "name": { "type": "string" },
    "last_name": { "type": "string" }
  },
  "required": []
}
```

2. When the schema does not define `required`, the validator should treat all defined properties as required.
```jsonc
// the validator should treat all defined properties as required
{
  "type": "object",
  "properties": {
    "name": { "type": "string" },
    "last_name": { "type": "string" }
  }
}
```

**Comment by braindeaf** [comment]
I encountered a problem this week. It was using RSwag / RSpec to test API responses against out swagger schema. We use strict mode, and this is handy I hadn't written the schema properly but I knew the response was correct. I worked my way through every error to get the swagger schema I wanted. 

Next step, we use `openapi-typescript` to generate Typescript from our swagger.json. Problem is in JSON-Schema we use strict and there is no equivalent in `openapi-typescript` and every single property is marked as optional by default. So then I have to go in and marked every property as required anyway just so the Typescript is correct. All of this because as I understand it the standard does not support a strict mode. So really is the answer to this not to have a strict mode at all and adhere to the JSON Schema standard and be done with it.

In order to be happy that the schema I am testing in RSwag is going to eventually generate the Typescript we need then we should write our specs without strict mode.

Just my two cents after two days looking into this (not 10 years 😸 )
